### PR TITLE
feat(core): add support for ambient directives

### PR DESCRIPTION
**Pull request by vsavkin**
Ambient directives can be configured when bootstraping an application.
Ambient directives can be used in every component of the application without
needing to explicitly list them.

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
Yes! I know that it can be abused by listing all the directives (thus effectively making them all "global") but we need _some_ kind of solution here, so let's try this one out.

File: modules/angular2/test/public_api_spec.ts [hunk 1]
```diff
@@ -398,8 +399,7 @@ var NG_API = [
   'CyclicDependencyError.message',
   'CyclicDependencyError.message=',
   'CyclicDependencyError.stackTrace',
-  'DEFAULT_PIPES',
-  'DEFAULT_PIPES_TOKEN',
+  'COMMON_PIPES',
```
Comment (IgorMinar) [inline modules/angular2/test/public_api_spec.ts]: what about "AMBIENT_*"?

File: modules/angular2/test/core/linker/integration_spec.ts [hunk 2]
```diff
@@ -1867,8 +1869,12 @@ class PushCmpWithRef {
   propagate() { this.ref.markForCheck(); }
 }
 
-@Component({selector: 'push-cmp-with-async', changeDetection: ChangeDetectionStrategy.OnPush})
-@View({template: '{{field | async}}'})
+@Component({
+  selector: 'push-cmp-with-async',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  template: '{{field | async}}',
+  pipes: [AsyncPipe]
```
Comment (IgorMinar) [inline modules/angular2/test/core/linker/integration_spec.ts]: why this change? isn't asyncpipe ambient?

File: modules/angular2/common.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * ```typescript
+ * import {NgClass, NgIf, NgFor, NgSwitch, NgSwitchWhen, NgSwitchDefault, NgModel, NgForm} from
+ * 'angular2/angular2';
+ * import {OtherDirective} from './myDirectives';
+ *
+ * @Component({
+ *   selector: 'my-component',
+ *   templateUrl: 'myComponent.html',
+ *   directives: [NgClass, NgIf, NgFor, NgSwitch, NgSwitchWhen, NgSwitchDefault, NgModel, NgForm]
```
Comment (gkalpak) [inline modules/angular2/common.ts]: Not exactly relevant, but `OtherDirective` is missing.

File: modules/angular2/src/core/compiler/ambient.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import {OpaqueToken} from "angular2/src/core/di";
+import {CONST_EXPR} from "angular2/src/core/facade/lang";
+
+/**
+ * A token that can be provided when bootstraping an application to make an array of directives
+ * available in every component of the application.
+ *
+ * ### Example
+ *
+ * Instead of writing:
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/ambient.ts]: The "instead of" section is actually missing.

File: modules/angular2/src/core/compiler/ambient.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+ */
+export const AMBIENT_DIRECTIVES = CONST_EXPR(new OpaqueToken("Ambient Directives"));
+
+/**
+ * A token that can be provided when bootstraping an application to make an array of pipes
+ * available in every component of the application.
+ *
+ * ### Example
+ *
+ * Instead of writing:
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/ambient.ts]: Same as above (this is not the "instead of" section).

File: modules/angular2/src/core/compiler/ambient.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ *   `
+ * })
+ * export class MyComponent {
+ *   ...
+ * }
+ *
+ * bootstrap(MyComponent, [provide(AMBIENT_DIRECTIVES, {useValue: [OtherDirective], multi:true})]);
+ * ```
+ */
+export const AMBIENT_DIRECTIVES = CONST_EXPR(new OpaqueToken("Ambient Directives"));
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/ambient.ts]: `OpaqueToken` type is missing.

BTW, since these are tokens, wouldn't it be more clear to call them `AMBIENT_<XYZ>_TOKEN` ? (Same as `DEFAULT_PIPES_TOKEN`.)
It doesn't read very well, but by importing `AMBIENT_PIPES`, I would expect to get some actual pipes.

File: modules/angular2/src/core/compiler/runtime_metadata.ts [hunk 7]
```diff
@@ -86,18 +88,23 @@ function removeDuplicates(items: any[]): any[] {
   return MapWrapper.keys(m);
 }
 
-function flattenDirectives(view: ViewMetadata): Type[] {
+function flattenDirectives(view: ViewMetadata, ambientDirectives: any[]): Type[] {
   if (isBlank(view.directives)) return [];
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/runtime_metadata.ts]: Do we still want to return `[]` ? What if there are `ambintDirectives` ?

File: modules/angular2/src/core/linker/proto_view_factory.ts [hunk 8]
```diff
@@ -118,9 +112,13 @@ export class ProtoViewFactory {
   }
 
   private _flattenPipes(view: ViewMetadata): any[] {
-    if (isBlank(view.pipes)) return this._defaultPipes;
-    var pipes = ListWrapper.clone(this._defaultPipes);
-    _flattenList(view.pipes, pipes);
+    let pipes = [];
+    if (isPresent(this._ambientPipes)) {
+      _flattenList(this._ambientPipes, pipes);
```
Comment (gkalpak) [inline modules/angular2/src/core/linker/proto_view_factory.ts]: Should this also be renamed to `_flattenArray` for consistency ?

File: modules/angular2/test/public_api_spec.ts (hunk 1 above)
Comment (vsavkin) [inline modules/angular2/test/public_api_spec.ts]: COMMON_ is a list of directives, and AMBIENT_ is the token.

File: modules/angular2/test/core/linker/integration_spec.ts (hunk 2 above)
Comment (vsavkin) [inline modules/angular2/test/core/linker/integration_spec.ts]: not in tests. I decided to remove ambient directives and pipes from tests to make them more precise.

File: modules/angular2/common.ts (hunk 3 above)
Comment (vsavkin) [inline modules/angular2/common.ts]: Fixed

File: modules/angular2/src/core/compiler/ambient.ts (hunk 6 above)
Comment (vsavkin) [inline modules/angular2/src/core/compiler/ambient.ts]: We removed the _TOKEN prefix. We can revisit it.

File: modules/angular2/src/core/compiler/runtime_metadata.ts (hunk 7 above)
Comment (vsavkin) [inline modules/angular2/src/core/compiler/runtime_metadata.ts]: Fixed

File: modules/angular2/src/core/linker/proto_view_factory.ts (hunk 8 above)
Comment (vsavkin) [inline modules/angular2/src/core/linker/proto_view_factory.ts]: Fixed

File: modules/angular2/test/public_api_spec.ts (hunk 1 above)
Comment (IgorMinar) [inline modules/angular2/test/public_api_spec.ts]: right, but isn't AMBIENT_ a public api? how will ionic add their stuff to the ambient list?

File: modules/angular2/test/core/linker/integration_spec.ts (hunk 2 above)
Comment (IgorMinar) [inline modules/angular2/test/core/linker/integration_spec.ts]: good.

**Comment by IgorMinar** [comment]
please resolve https://github.com/angular/angular/pull/5075#discussion_r43702776, otherwise lgtm

File: modules/angular2/test/public_api_spec.ts [hunk 9]
```diff
@@ -398,8 +400,8 @@ var NG_API = [
...
+  'AMBIENT_PIPES:js',
```
Comment (gkalpak) [inline modules/angular2/test/public_api_spec.ts]: Probably not important, but this list used to be alphabetically sorted.

**Comment by gkalpak** [comment]
So, if I get this correctly, `COMMON_DIRECTIVES/PIPES` will be ambient by default.

I remember this was previously decided against (for performance reasons I think).
OOC, is it that the new offline compiler will remove the perf hit (so we can use ambient directives/pipes, for improving developer ergonomics) ?

**Comment by pkozlowski-opensource** [comment]
> I remember this was previously decided against (for performance reasons I think).

There were discussions, but perf wasn't the factor (it is compile time-thing so we are not on a critical path here).

**Comment by pkozlowski-opensource** [comment]
The main argument against doing this was template "encapsulation": say we one day add a new thing to core directives / pipes and this might conflict with an existing directives / might start to activate on parts of HTML that were not active before.

**Comment by gkalpak** [comment]
I see. Thx @pkozlowski-opensource !
